models/classes/StatisticalMetadata/Import/Builder/ReportBuilder.php

```diff
@@ -0,0 +1,247 @@
+use oat\tao\model\StatisticalMetadata\Import\Exception\HeaderValidationException;
+use oat\tao\model\StatisticalMetadata\Import\Exception\WarningValidationException;
+use oat\tao\model\StatisticalMetadata\Import\Exception\AbstractValidationException;
+
+class ReportBuilder
+{
+    public function buildByReporter(ImportReporter $reporter): Report
+    {
+        $reportType = $this->getReportType($reporter);
+
+        $subReport = $this->createReportByResults($reportType, $reporter);
+        $report = $this
+            ->createReportByResults($reportType, $reporter)
+            ->add($subReport);
+
+        $onlyWarningReports = [];
+        $warningAndErrorReports = [];
+
+        /** @var WarningValidationException[]|ErrorValidationException[] $exceptions */
+        foreach ($reporter->getWarningsAndErrors() as $line => $exceptions) {
+            $headerExceptionReports = [];
+            $warningReports = [];
+            $errorReports = [];
+
+            foreach ($exceptions as $exception) {
+                if ($exception instanceof HeaderValidationException) {
+                    $headerExceptionReports[] = $this->createReportByException($exception);
+
+                    continue;
+                }
+
+                if ($exception instanceof WarningValidationException) {
+                    $warningReports[] = $this->createReportByException($exception);
+
+                    continue;
+                }
+
+                if ($exception instanceof ErrorValidationException) {
+                    $errorReports[] = $this->createReportByException($exception);
+                }
+            }
+
+            if (empty($headerExceptionReports) && empty($errorReports) && empty($warningReports)) {
+                continue;
+            }
+
+            $lineMainReportType = empty($headerExceptionReports) && empty($errorReports)
+                ? Report::TYPE_WARNING
+                : Report::TYPE_ERROR;
+            $lineMainReport = Report::create($lineMainReportType, 'line %s: %s', [$line, '']);
+
+            foreach ($headerExceptionReports as $headerExceptionReport) {
+                $lineMainReport->add($headerExceptionReport);
+            }
+
+            foreach ($warningReports as $warningReport) {
+                $lineMainReport->add($warningReport);
+            }
+
+            foreach ($errorReports as $errorReport) {
+                $lineMainReport->add($errorReport);
+            }
+
+            if (empty($errorReports)) {
+                $onlyWarningReports[$line] = $lineMainReport;
+            }
+
+            if (!empty($errorReports)) {
+                $warningAndErrorReports[$line] = $lineMainReport;
+            }
+        }
+
+        if (!empty($headerExceptionReports)) {
+            $this->createAndAddSubReport(
+                $report,
+                $headerExceptionReports,
+                Report::TYPE_ERROR,
+                'Header contain an error(s)'
```

Comment:
I am not sure `an` will match with both plural an singular

```suggestion
                'Header contain error(s)'
```

Reply:
Same for other messages

Reply:
Maybe a ternary like `count($headerExceptionReports) > 1 ? 'Header contains an error' : 'Header contains errors'` would work in these cases.